python_ta/checkers/type_annotation_checker.py

```diff
@@ -0,0 +1,57 @@
+"""checker for type annotation.
+"""
+
+import astroid
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+
+
+class TypeAnnotationChecker(BaseChecker):
+
+    __implements__ = IAstroidChecker
+
+    name = 'TypeAnnotationChecker'
+    msgs = {'E9970': ('Function argument is missing type annotation',
+                      'type-annotation-arg',
+                      'Presented when a type annotation is missing.'),
+            'E9971': ('Function is missing return type annotation',
+                      'type-annotation-return',
+                      'Presented when a type annotation is missing.'),
+            'E9972': ('Class variable is missing type annotation',
+                      'type-annotation-class-var',
+                      'Presented when a type annotation is missing.'),
+            'E9973': ('Instance variable is missing type annotation',
+                      'type-annotation-inst-var',
+                      'Presented when a type annotation is missing.'),
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    def visit_functiondef(self, node):
+        if isinstance(node.parent, astroid.ClassDef):
+            # Function is a method. Type annotation not required for first argument 'self'
+            if any(arg is None for arg in node.args.annotations[1:]):
+                self.add_message('type-annotation-arg', node=node)
+            # __init__ has return type None
+            if node.returns is None and node.name != '__init__':
```

Comment:
This is a bit scary. Are you saying that astroid `returns` attribute can't distinguish between the following cases:

```
def f():

def f() -> None
```

If so, there's a larger problem here.

Reply:
In the second case, the `node.returns` attribute would contain a `Const` node with the value None, so it should be able to distinguish between the two cases

Reply:
Okay, then this is not a special case. Students are expected to annotate the return type of `__init__`.